### PR TITLE
🐛 dependency_hash

### LINKED ISSUE #47: lockfile check not running
I need to confirm but I believe the latest version of hatch is skipping checks with its new use of [dependency_hash](https://hatch.pypa.io/latest/plugins/environment/reference/#hatch.env.plugin.interface.EnvironmentInterface.dependency_hash) (https://hatch.pypa.io/latest/blog/2023/12/11/hatch-v180/#faster-environment-usage) - this allows lockfiles to be out of date (by hand editing at least)

> Faster environment usage
> [Spawning a shell](https://hatch.pypa.io/latest/environment/#entering-environments) or [running commands](https://hatch.pypa.io/latest/environment/#command-execution) within environments always first checks that your project's dependencies are satisfied and if not synchronizes the environment with what is defined. Previously, this had the potential to be quite slow for projects that have many dependencies.
> Now the set of dependency definitions is [hashed](https://hatch.pypa.io/latest/plugins/environment/reference/#hatch.env.plugin.interface.EnvironmentInterface.dependency_hash) and no check is performed if the hash is the same as before, significantly speeding up environment usage in most cases.

https://github.com/pypa/hatch/blob/d3246e957584d292319e7b93301598cdf611e902/src/hatch/cli/application.py#L107-L117

### PR DESCRIPTION
**Pull request by juftin**
Resolves #47

### `dependency_hash`

This PR implements `dependency_hash` function on the plugin which is required for `hatch >= 1.8.0`. Before this fix the new `dependency_hash` behavior allows for lockfiles to be out of date without getting updated because their `dependency_hash` still matches.

Currently, `dependency_hash` calls `run_pip_compile` which regenerates the lockfile when needed, it then hashes `super().dependency_hash()` + the hash of the lockfile contents.

```python
def dependency_hash(self) -> str:
    """
    Get the dependency hash
    """
    self.run_pip_compile()
    hatch_hash = super().dependency_hash()
    if not self.dependencies:
        return hatch_hash
    else:
        lockfile_hash = self.piptools_lock.get_hash()
        return hashlib.sha256(f"{hatch_hash}-{lockfile_hash}".encode()).hexdigest()
```

#### Integration Tests

- New tests including the click `CliRunner` which allows for testing `hatch` CLI command with the plugin
- More test cases for existing integration tests

#### Testing Environements

The project now includes a testing matrix where hatch `1.7.x`, `1.8.x`, and `1.9.x` are all tested in addition to Python `3.8` though `3.12`.

### Version Pinning

After the latest tests, this plugin is found to be compatible with `hatch>=1.7.0`

### Lock Files

Due to above mentioned version pinning all lockfiles were regenerated - I used macOS to generate these lockfiles so some linux specific dependencies will be removed. 

### `plugin_check_command`

Wrapped with `safe_activation` to ensure it does the right thing wherever it's used

### CODE REVIEW

File: hatch_pip_compile/plugin.py [hunk 1]
```diff
@@ -90,6 +90,14 @@ def get_option_types() -> Dict[str, Any]:
             "pip-compile-install-args": List[str],
         }
 
+    def dependency_hash(self) -> str:
+        """
+        Get the dependency hash
+        """
+        if not self.lockfile_up_to_date:
+            self.sync_dependencies()
+        return super().dependency_hash()
```
Comment (juftin) [inline hatch_pip_compile/plugin.py:99]: I don't like calling `lockfile_up_to_date` / `sync_dependencies` here, but I'm not sure there's a way around it

Comment (juftin) [reply hatch_pip_compile/plugin.py:99]: https://github.com/pypa/hatch/blob/d3246e957584d292319e7b93301598cdf611e902/src/hatch/cli/application.py#L107-L117

Hatch's workflow:
1. Check current `dependency_hash`
2. Compare it to stored hash
3. If they don't match then run `dependencies_in_sync` / `sync_dependencies` 
4. Update the stored hash with hash from step 1

Comment (juftin) [reply hatch_pip_compile/plugin.py:99]: This eliminates the need for `dependencies_in_sync` / `sync_dependencies` to be called after `dependency_hash`

I've opened https://github.com/pypa/hatch/issues/1168 which I think would make all of this a lot cleaner

File: hatch_pip_compile/plugin.py [hunk 2]
```diff
@@ -90,6 +90,13 @@ def get_option_types() -> Dict[str, Any]:
             "pip-compile-install-args": List[str],
         }
 
+    def dependency_hash(self) -> str:
+        """
+        Get the dependency hash
+        """
+        self.run_pip_compile()
+        return super().dependency_hash()
```
Comment (juftin) [inline hatch_pip_compile/plugin.py:98]: `run_pip_compile()` checks whether the lockfile is up to date before proceeding

**Comment by juftin** [comment]
Hey @oprypin could I get you to take a look at this and let me know your thoughts? 

My long term strategy is for https://github.com/pypa/hatch/pull/1169 to be accepted which would allow me to return a completely random hash from `dependency_hash` when the lockfile is out of sync and then return a deterministic hash when the lockfile _is_ in sync. Until that's implemented I don't see a way of getting around this except for calling `sync_dependencies` up front along with `dependency_hash`. 

The workflow is the essentially the same as before except `sync_dependencies` is never called a second time because it's guarded by `dependencies_in_sync`. That fix mentioned above would allow `dependency_hash` / `dependencies_in_sync` / `sync_dependencies` to all perform their proper job.

**Review by oprypin — APPROVED** [review]
I haven't looked *too* carefully, but all solutions applied here make sense

**Pull request merged**